### PR TITLE
User auth  adva

### PR DESCRIPTION
**Pull request by AtomicManic**

### CODE REVIEW

File: controller/external.controller.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+const {
+  getUserByUuid,
+  addUser,
+  insertExperiment,
+  getUserExperiment,
+} = require("./user.controller");
+const { checkAttributes } = require("./../Service/utils");
+
+const runTest = async (req, res, next) => {
```
Comment (sufka861) [inline controller/external.controller.js:9]: great abstraction of all of the logic

File: controller/user.controller.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  if (!newUser) throw new ServerUnableError("create");
+  return newUser;
+};
+
+const insertExperiment = async (uuid, experiment) => {
+  if (!uuid) throw new PropertyNotFound("uuid");
+  if (!experiment.variant) throw new PropertyNotFound("variant");
+  const response = await userRepository.retrieveByUuid(uuid);
+  const user = response[0];
+  console.log(user);
```
Comment (sufka861) [inline controller/user.controller.js:36]: no need for log

File: middleware/errorHandler.mw.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 const logger = require("./../middleware/loggers/error.logger");
 
 exports.errorHandler = (error, req, res, next) => {
+  console.log(error);
```
Comment (sufka861) [inline middleware/errorHandler.mw.js:4]: no need for logger here

File: models/user.model.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const { ObjectId } = require("bson");
+const { Schema, model } = require("mongoose");
+
+const userSchema = new Schema(
+  {
+    uuid: { type: String, required: true },
+    experiments: [
+      {
+        experimentId: String,
```
Comment (sufka861) [inline models/user.model.js:9]: can add validation on valid experiment Id of type ObjectId

File: models/user.model.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+const { ObjectId } = require("bson");
+const { Schema, model } = require("mongoose");
+
+const userSchema = new Schema(
+  {
+    uuid: { type: String, required: true },
```
Comment (sufka861) [inline models/user.model.js:6]: can add validation on valid uuid

File: repositories/user.repository.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+const MongoStorage = require("../db/mongo.storage");
+
+module.exports = new (class UsersRepository extends MongoStorage {
+  constructor() {
+    super("user");
+  }
+
+  find() {
```
Comment (peerfikhman) [inline repositories/user.repository.js:8]: redandent becouse of heritage

File: router/user.routes.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  checkAttributes,
+  getUserByUuid,
+  setCookie,
+  getCookie,
+  addUser,
+  insertExperiment,
+  getUserExperiment,
+} = require("../controller/user.controller");
+
+// Get user by uuid
```
Comment (sufka861) [inline router/user.routes.js:13]: no need for comments

File: router/user.routes.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+// Get cookie values
+userRouter.get("/cookie", getCookie);
+
+// Create new user
+userRouter.post("/", addUser);
+
+// Add experiment to user
+userRouter.put("/:uuid", insertExperiment);
+
+module.exports = { userRouter };
```
Comment (sufka861) [inline router/user.routes.js:31]: needed only for inside testing - can be deleted in master

**Review by sufka861 — APPROVED** [review]
view and correct comments below - otherwise approved good job ! :)

**Pull request closed without merging**